File: Codechef/2021/Codechef_JUN_LunchTime_2021/Tic_Tac_Toe.py
test = int(input())
for _ in range(test):
    n, m, k = map(int, input().split())
    turn = k * k * 2
    alice = [[0] * m] * n
    bob = [[0] * m] * n

    flag = False
    if turn > n * m:
        print("Draw")
        continue
    else:
        for temp in range(1, n * m + 1):
            x, y = map(int, input().split())
            x -= 1
            y -= 1
            print(bob)
            if temp % 2 == 1:
                alice[x][y] = 1
            else:   
                print(x, y)
                bob[x][y] = 1

            if temp >= turn:
                check = [[0] * n] * m

                for i in range(n):
                    check[i][0] = alice[i][0]
                for j in range(m):
                    check[0][j] = alice[0][j]
                for i in range(1, n):
                    for j in range(1,m):
                        if alice[i][j] == 1:
                            check[i][j] = min(check[i - 1][j - 1], check[i - 1][j], check[i][j - 1]) + 1
                        else:
                            check[i][j] = 0
                        if check[i][j] == k:
                            print("Alice")
                            flag = True
                        if flag:
                            break
                    if flag:
                        break
                if flag:
                    break

                check = [[0] * n] * m

                for i in range(n):
                    check[i][0] = bob[i][0]
                for j in range(m):
                    check[0][j] = bob[0][j]
                for i in range(1, n):
                    for j in range(1,m):
                        if bob[i][j] == 1:
                            check[i][j] = min(check[i - 1][j - 1], check[i - 1][j], check[i][j - 1]) + 1
                        else:
                            check[i][j] = 0
                        if check[i][j] == k:
                            print("Bob")
                            flag = True
                        if flag:
                            break
                    if flag:
                        break
                if flag:
                    break
        if flag:
            pass
        else:
            print("Draw")

File: Codechef/2021/Codechef_JUN_LunchTime_2021/Tic_Tac_Toe.cpp
#include <bits/stdc++.h>
#define ll long long int
#define endl "\n"
using namespace std;

int main()
{   
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int test;
    cin >> test;
    while(test--)
    {
        int i, j, k, n, temp, count = 0, ans = 0, sum = 0;
        int m;
        cin >> n >> m >> k;
        
        swap(n, m);
        int turn = k * k * 2;
        int alice[n][m], bob[n][m];

        for(i = 0; i < n; i++)
        {
            for(j = 0; j < n; j++)
            {
                alice[i][j] = 0;
                bob[i][j] = 0;
            }
        }

        if(turn > n * m)
        {
            cout << "Draw" << endl;
            continue;
        }
        else
        {
            for(temp = 0; temp < n * m; temp++)
            {   
                cin >> i >> j;
                i--;j--;
                if(temp % 2 == 0)
                {
                    alice[i][j] = 1;
                }
                else
                {
                    bob[i][j] = 1;
                }

                if(temp >= turn)
                {
                    int check[n][m];

                    /*Alice check*/
                    for(i = 0; i < n; i++)
                    {
                        check[i][0] = alice[i][0];
                    }
                    for(j = 0; j < m; j++)
                    {
                        check[0][j] = alice[0][j];
                    }
                    for(i = 1; i < n; i++)
                    {
                        for(j = 1; j < n; j++)
                        {
                            if(alice[i][j] == 1)
                            {
                                check[i][j] = min(check[i - 1][j - 1], min(check[i - 1][j], check[i][j - 1])) + 1;
                            }
                            else
                            {
                                check[i][j] = 0;
                            }
                        }
                    }

                    for(i = 0; i < n; i++)
                    {
                        for(j = 0; j < n; j++)
                        {
                            if(check[i][j] == k)
                            {
                                ans = 1;
                                break;
                            }
                        }
                        if(ans)break;
                    }

                    if(ans)break;

                    /*Bob check*/
                    for(i = 0; i < n; i++)
                    {
                        check[i][0] = bob[i][0];
                    }
                    for(j = 0; j < m; j++)
                    {
                        check[0][j] = bob[0][j];
                    }
                    for(i = 1; i < n; i++)
                    {
                        for(j = 1; j < n; j++)
                        {
                            if(bob[i][j] == 1)
                            {
                                check[i][j] = min(check[i - 1][j - 1], min(check[i - 1][j], check[i][j - 1])) + 1;
                            }
                            else
                            {
                                check[i][j] = 0;
                            }
                        }
                    }

                    for(i = 0; i < n; i++)
                    {
                        for(j = 0; j < n; j++)
                        {
                            if(check[i][j] == k)
                            {
                                ans = 2;
                                break;
                            }
                        }
                        if(ans)break;
                    }

                    if(ans)break;
                }
                if(ans)break;
            }
            if(ans == 1)
            {
                cout << "Alice" << endl;
            }
            else if(ans == 2)
            {
                cout << "Bob" << endl;
            }
            else
            {
                cout << "Draw" << endl;
            }
        }
    }

return 0;
}



